### PR TITLE
feat: support aria-disabled and disabled for menu and listbox

### PR DESCRIPTION
**Pull request by LuLaValva**
- Fixed https://github.com/makeup/makeup-js/issues/87
- ebay/ebayui-core#1757
- ebay/ebayui-core#1693

# Changes
- Items in `makeup-listbox` and `makeup-menu` may no longer be interacted with when `aria-disabled="true"` or `disabled`.
- Navigation emitter skips past disabled items when using arrow keys.

### CODE REVIEW

**Comment by LuLaValva** [comment]
> Not looked through it in detail yet, but could we not just add `disabled` and `aria-disabled=true` to the filtering function?
> 
> https://github.com/makeup/makeup-js/blob/master/packages/makeup-navigation-emitter/src/index.js#L17
> 
> Then update the mutation observer filter too:
> 
> https://github.com/makeup/makeup-js/blob/master/packages/makeup-navigation-emitter/src/index.js#L216

I explored this avenue before moving on to my method, but it was causing problems with indexing so I figured it was better to separate hidden items and disabled ones. I will explore it further now that I know more about how the `makeup-navigation-emitter` works, it may lead to a cleaner solution.

**Comment by ianmcburnie** [comment]
You probably ran into indexing issues due to the mutationObserver not firing for hidden and aria-hidden, which would mean the model wouldn't get updated:

https://github.com/makeup/makeup-js/blob/master/packages/makeup-navigation-emitter/src/index.js#L84

**Comment by ianmcburnie** [comment]
@LuLaValva I had a little more time to look into this, and I now understand the direction you took (with the help of your explanation offline too). Hidden items _do_ need to be treated differently from disabled items (which is probably why I didn't include disabled items in the filter in the first place). To avoid future confusion it might be best if we rename `filteredItems` to `visibleItems` and maybe just inline the `itemFilter` function?

**Comment by LuLaValva** [comment]
After some more discussion offline and the discovery that the `hidden` attribute did not work properly with `makeup-menu` and `makeup-listbox`, I have opted for the following:
- Generalize the notion of ignoring an option by attributes, and provide it as a config option (`ignoreByAttrs`)
- Filter the list inside `makeup-navigation-emitter` based on `ignoreByAttrs`, which may be controlled externally
  - This has the effect of only placing `data-makeup-index` on items which are _not_ ignored
- Inside `makeup-menu` and `makeup-listbox`, void clicks and button presses that target elements without `data-makeup-index`

**Comment by ianmcburnie** [comment]
Can we have the style/syntax changes pushed out to #93? That will help keep the noise down for reviewers.

**Comment by LuLaValva** [comment]
> Can we have the style/syntax changes pushed out to #93? That will help keep the noise down for reviewers.

Yep! I meant to keep them all out of this PR, but it looks like a few slipped through the cracks. They're removed now.

File: packages/makeup-active-descendant/dist/cjs/index.js [hunk 1]
```diff
@@ -45,7 +45,12 @@ var defaultOptions = {
   autoReset: -1,
   autoScroll: false,
   axis: 'both',
-  ignoreButtons: false
+  ignoreButtons: false,
+
+  /** @type {{[attr: string]: unknown}} */
```
Comment (ianmcburnie) [inline packages/makeup-active-descendant/dist/cjs/index.js:50]: Just confirming, are these annotations following this spec: https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html or something else?

Comment (LuLaValva) [reply packages/makeup-active-descendant/dist/cjs/index.js:50]: These are following the typescript-based JSDoc spec, yes

**Review by ianmcburnie — CHANGES_REQUESTED** [review]
Couple of things I noticed quickly:

1. Without making any changes to config/options, I changed item 1 to aria-disabled="true" in the example markup and when navigating with keyboard the example gets out of sync. See both item 1 and item 3 have dotted lines. It seems a class isn't being removed.


Screen Shot 2022-09-28 at 11 57 07 AM

Screen Shot 2022-09-28 at 11 57 32 AM

Possibly related to this area of logic: https://github.com/makeup/makeup-js/pull/88/files#diff-a9b0afebe911791425e03a9d82102e1b68d37abd445a7cca0a2d0956c580c591R40

2. Can you please update options sections of README to document usage of ignoreByAttrs

File: packages/makeup-active-descendant/dist/cjs/index.js (hunk 1 above)
Comment (ianmcburnie) [reply packages/makeup-active-descendant/dist/cjs/index.js:50]: Thanks!

**Review by ianmcburnie — CHANGES_REQUESTED** [review]
Ran into this issue when applying `hidden` attribute to "Item 1" in navigation emitter demo page. When using the up down arrow keys it is cycling through 0, 1, and 2. I would expect it only to be 0 and 1 due to the hidden item. I confirmed this issue doesn't happen on master.

Screen Shot 2022-09-28 at 8 23 30 PM

**Comment by LuLaValva** [comment]
Good catch @ianmcburnie! I did not properly understand how default attribute values were assigned and had to change my comparison to treat booleans differently than strings. Should be resolved now.

**Comment by agliga** [comment]
Couple of things I noticed.

1. We should probably add some examples of disabled states in the docs. Probably under menu listbox and maybe roving tabindex. This is helpful so we can test it and any further changes.
2. I noticed in menu there's a small bug. I tabbed through and saw that it actually tabbed through the first item when it was disabled. (Again in this example I'm hitting tab only. It tabs through the first item, then the second... If I turn off disable then it correctly tabs through only the first item)
drwer

Otherwise it seems to be working great!

**Comment by LuLaValva** [comment]
Thanks @agliga, this was a tricky one but it should be fixed now. I also added an extra bit of documentation to `makeup-menu`, if you'd like to see more let me know.

File: packages/makeup-active-descendant/README.md [hunk 2]
```diff
@@ -158,6 +158,17 @@ Use CSS to style the active descendant however you wish:
 * `autoScroll` : Specify true to scroll the container as activeDescendant changes (default: false)
 * `axis` : specify 'x' for left/right arrow keys, 'y' for up/down arrow keys, or 'both' (default: 'both')
 * `ignoreButtons`: if set to true, nested button elements will not trigger navigationModelChange events. This is useful in a combobox + button scenario, where only the textbox should trigger navigationModelChange events (default: false)
+* `ignoreByAttrs`: Map from attribute names to the values that are associated with ignoring those names
```
Comment (ianmcburnie) [inline packages/makeup-active-descendant/README.md:161]: Can you please add the same to the READMEs of roving-tabindex and navigation-emitter. Thanks!

**Comment by ianmcburnie** [comment]
Thinking about this more, the main nitpick I have is that previously the `data-makeup-index` would be ever-present on every element that matched the `itemSelector`.  As it stands now, if we have 1000 items and we change the first item to be disabled, then all 1000 dom elements are touched when having their `data-makeup-index` updated.

**Comment by ianmcburnie** [comment]
> Thinking about this more, the main nitpick I have is that previously the `data-makeup-index` would be ever-present on every element that matched the `itemSelector`. As it stands now, if we have 1000 items and we change the first item to be disabled, then all 1000 dom elements are touched when having their `data-makeup-index` updated.

Having said that, `hidden` has the same existing "issue". Although the use case for toggling aria-disabled is probably more common than hidden.

Anyhow, I also have a couple of ideas for improving the pre-existing code. A few things that should make everything a little easier to follow. I'll push those to a different branch and sync up with you when done so we can compare notes.

**Pull request closed without merging**